xdsl/pattern_rewriter.py

```diff
@@ -318,11 +319,19 @@ class RewritePattern(ABC):
     A side-effect free rewrite pattern matching on a DAG.
     """
 
+    traits: ClassVar[tuple[type[Trait], ...]] = ()
+
+    def _match_and_rewrite(self, op: Operation,
+                           rewriter: PatternRewriter) -> None:
+        if all(isinstance(op, trait) for trait in self.traits):
```

Comment:
I am not sure if this implicit `AND` of all given traits is the best idea, it might be confusing that `traits = (Pure, SomeOtherLessStrictPure)` does not work because an operation needs to satisfy both and not just one of them. I think we should try and be as specific as possible with this interface.

Reply:
I'm confused, you mean to say that we need better documentation of the behaviour?

Reply:
I'm not sure if a list of traits is the right design here, as it introduces an implicit "and" relation between them, where others might expect an "or".

We should either make the "and" explicit, only allow matching a single trait, or add a way to select and/or behaviour.

Maybe we can rename the attribute `matching_all_traits = (Trait1, Trait2)` (where the "and" is more explicit) or something similar?